kong/clustering/rpc/manager.lua

```diff
@@ -0,0 +1,365 @@
+  end
+end
+
+
+-- Helper that finds a node by node_id and check
+-- if capability is supported
+-- Returns: "local" if found locally,
+-- or "concentrator" if found from the concentrator
+-- In case of error, return nil, err instead
+function _M:_find_node_and_check_capability(node_id, cap)
+  if self.client_capabilities[node_id] then
+    if not self.client_capabilities[node_id].set[cap] then
+      return nil, "requested capability does not exist, capability: " ..
+                  cap .. ", node_id: " .. node_id
+    end
+
+    return "local"
+  end
+
+  -- does concentrator knows more about this client?
+  local res, err = kong.db.clustering_data_planes:select({ id = node_id })
+  if err then
+    return nil, "unable to query concentrator " .. err
+  end
+
+  if not res or ngx_time() - res.last_seen > CLUSTERING_PING_INTERVAL * 2 then
+    return nil, "node is not connected, node_id: " .. node_id
+  end
+
+  for _, c in ipairs(res.rpc_capabilities) do
+    if c == cap then
+      return "concentrator"
+    end
+  end
+
+  return nil, "requested capability does not exist, capability: " ..
+              cap .. ", node_id: " .. node_id
+end
+
+
+-- low level helper used internally by :call() and concentrator
+-- this one does not consider forwarding using concentrator
+-- when node does not exist
+function _M:_local_call(node_id, method, params)
+  if not self.client_capabilities[node_id] then
+    return nil, "node is not connected, node_id: " .. node_id
+  end
+
+  local cap = utils.parse_method_name(method)
+  if not self.client_capabilities[node_id].set[cap] then
+    return nil, "requested capability does not exist, capability: " ..
+                cap .. ", node_id: " .. node_id
+  end
+
+  local s = next(self.clients[node_id])
+
+  local fut = future.new(node_id, s, method, params)
+  assert(fut:start())
+
+  local ok, err = fut:wait(5)
+  if err then
+    return nil, err
+  end
+
+  if ok then
+    return fut.result
+  end
+
+  return nil, fut.error.message
+end
+
+
+-- public interface, try call on node_id locally first,
+-- if node is not connected, try concentrator next
+function _M:call(node_id, method, ...)
+  local cap = utils.parse_method_name(method)
+
+  local res, err = self:_find_node_and_check_capability(node_id, cap)
+  if not res then
+    return nil, err
+  end
+
+  local params = {...}
+
+  if res == "local" then
+    res, err = self:_local_call(node_id, method, params)
+    if not res then
+      return nil, err
+    end
+
+    return res
+  end
+
+  assert(res == "concentrator")
+
+  -- try concentrator
+  local fut = future.new(node_id, self.concentrator, method, params)
+  assert(fut:start())
+
+  local ok, err = fut:wait(5)
+  if err then
+    return nil, err
+  end
+
+  if ok then
+    return fut.result
+  end
+
+  return nil, fut.error.message
+end
+
+
+-- handle incoming client connections
+function _M:handle_websocket()
+  local kong_version = ngx_var.http_x_kong_version
```

Comment:
Should we use `tools.get_header()` to ensure we get the first header's value?

Reply:
This is probably unnecessary, as the header is sent by DP code, and receiving multiple `X-Kong-Version` header will constitute a bug.